File: packages/identity-py/fuzefront_identity/graph_create.py
"""Graph create: ``lid`` in, ``idMap`` out.

Semantics mirror ``shared/src/identity/graph-create.ts`` exactly, including the
error codes, so a graph accepted by a Node service is accepted identically by a
Python one.

Each node carries a document-scoped local id (``lid``) and references it as
``"lid:<local>"`` — JSON:API 1.1 ``lid``, SCIM ``bulkId`` (RFC 7644 3.7.2) and
OData ``$batch`` Content-ID all solve this the same way. Deliberately NOT the
``id`` field: a placeholder there would put ``id`` back in create bodies and
degrade the rule to "id must be present but fake", which is unenforceable.

Every id is minted UP FRONT and substituted in, so handlers never learn ``lid``
existed, ``idMap`` is just the allocation table, and reference cycles resolve
without a deferred second write.
"""

from __future__ import annotations

import re
from typing import Any, Dict, Iterable, Optional, Set, Tuple

from .ids import mint_id
from .registry import is_entity_type

_LID_REF = re.compile(r"^lid:(.+)$")

# Permitted shape of a local id, matching the ``LocalId`` schema published in the
# contract (1-64 chars) and the TypeScript LID_FORMAT. Enforced because ``lid``
# values are echoed back as ``idMap`` KEYS — the one place client-controlled text
# crosses into a response — and because a contract that declares a bound the
# implementation does not enforce is not a bound at all.
_LID_FORMAT = re.compile(r"^[A-Za-z0-9_.:-]{1,64}$")

DEFAULT_MAX_NODES = 500
DEFAULT_MAX_DEPTH = 32

MUTATING_METHODS = frozenset({"POST", "PUT", "PATCH"})


class GraphCreateError(ValueError):
    """Raised when a create graph is malformed or violates the standard."""

    def __init__(self, code: str, path: str, message: str) -> None:
        super().__init__(message)
        self.code = code
        self.path = path

    def as_response(self) -> Dict[str, Any]:
        """The canonical 422 body, identical to the Node middleware's."""
        return {
            "error": "unprocessable_entity",
            "code": self.code,
            "message": str(self),
            "path": self.path,
        }


def resolve_graph(
    body: Any,
    aggregate: Iterable[str],
    max_nodes: int = DEFAULT_MAX_NODES,
    max_depth: int = DEFAULT_MAX_DEPTH,
) -> Tuple[Any, Dict[str, str]]:
    """Validate a create graph, mint an id per ``lid`` node, resolve references.

    ``aggregate`` is the set of entity types this service OWNS. A ``lid`` node
    declaring anything else is rejected: a graph spanning services cannot be
    created atomically, so cross-service entities must be referenced by their
    existing ids rather than created here.

    Returns ``(rewritten_body, id_map)``. Mutates nothing.
    """
    owned: Set[str] = set(aggregate)
    id_map: Dict[str, str] = {}
    node_count = 0

    def collect(value: Any, path: str, depth: int) -> None:
        nonlocal node_count
        if depth > max_depth:
            raise GraphCreateError("GRAPH_TOO_DEEP", path, f"graph exceeds {max_depth} levels")
        if isinstance(value, list):
            for index, item in enumerate(value):
                collect(item, f"{path}/{index}", depth + 1)
            return
        if not isinstance(value, dict):
            return

        if "id" in value:
            raise GraphCreateError(
                "CLIENT_SUPPLIED_ID",
                path,
                "ids are minted by the owning service and must not be supplied on create",
            )

        if "lid" in value:
            lid = value["lid"]
            if not isinstance(lid, str) or not _LID_FORMAT.match(lid):
                raise GraphCreateError(
                    "MALFORMED_LID", path, "lid must be 1-64 characters of [A-Za-z0-9_.:-]"
                )
            if lid in id_map:
                raise GraphCreateError("DUPLICATE_LID", path, f"lid {lid!r} is declared twice")
            node_count += 1
            if node_count > max_nodes:
                raise GraphCreateError("GRAPH_TOO_LARGE", path, f"graph exceeds {max_nodes} nodes")

            entity_type = value.get("type")
            if not isinstance(entity_type, str):
                raise GraphCreateError("MISSING_TYPE", path, "a lid node must declare its entity type")
            if not is_entity_type(entity_type):
                raise GraphCreateError("UNKNOWN_TYPE", path, f"unregistered entity type {entity_type!r}")
            if entity_type not in owned:
                raise GraphCreateError(
                    "CROSS_AGGREGATE_LID",
                    path,
                    f"{entity_type} is owned by another service; reference it by its "
                    "existing id instead of creating it here",
                )

            id_map[lid] = mint_id(entity_type)

        for key, child in value.items():
            collect(child, f"{path}/{key}", depth + 1)

    collect(body, "", 0)

    def rewrite(value: Any, path: str) -> Any:
        if isinstance(value, list):
            return [rewrite(item, f"{path}/{index}") for index, item in enumerate(value)]
        if isinstance(value, str):
            match = _LID_REF.match(value)
            if not match:
                return value
            target = match.group(1)
            resolved = id_map.get(target)
            if resolved is None:
                raise GraphCreateError(
                    "UNKNOWN_LID", path, f"reference to undeclared lid {target!r}"
                )
            return resolved
        if not isinstance(value, dict):
            return value

        out: Dict[str, Any] = {}
        for key, child in value.items():
            if key == "lid":
                continue  # document-scoped; never persisted
            out[key] = rewrite(child, f"{path}/{key}")
        lid = value.get("lid")
        if isinstance(lid, str):
            out["id"] = id_map[lid]
        return out

    return rewrite(body, ""), id_map


# ---------------------------------------------------------------------------
# ASGI middleware (FastAPI / Starlette)
# ---------------------------------------------------------------------------


class GraphCreateMiddleware:
    """Pure-ASGI middleware applying :func:`resolve_graph` to JSON create bodies.

    Deliberately stdlib-only — no Starlette import — so the package installs into
    any Python microservice regardless of framework, and works under FastAPI,
    Starlette or any bare ASGI app::

        app.add_middleware(GraphCreateMiddleware, aggregate={"customer", "invoice"})

    Handlers downstream receive a body whose ids are already minted and whose
    references are already resolved, and the ``idMap`` is merged into the JSON
    response automatically — so a route opts in by doing nothing at all.
    """

    def __init__(
        self,
        app: Any,
        aggregate: Iterable[str],
        max_nodes: int = DEFAULT_MAX_NODES,
        max_depth: int = DEFAULT_MAX_DEPTH,
    ) -> None:
        self.app = app
        self.aggregate = set(aggregate)
        self.max_nodes = max_nodes
        self.max_depth = max_depth

    async def __call__(self, scope: Dict[str, Any], receive: Any, send: Any) -> None:
        import json

        if scope.get("type") != "http" or scope.get("method", "").upper() not in MUTATING_METHODS:
            await self.app(scope, receive, send)
            return

        raw = await _read_body(receive)
        if not raw:
            await self.app(scope, _replay(raw), send)
            return

        try:
            parsed = json.loads(raw)
        except ValueError:
            # Not JSON — not ours to interpret; hand it on untouched.
            await self.app(scope, _replay(raw), send)
            return

        try:
            rewritten, id_map = resolve_graph(
                parsed, self.aggregate, self.max_nodes, self.max_depth
            )
        except GraphCreateError as err:
            await _send_json(send, 422, err.as_response())
            return

        new_body = json.dumps(rewritten).encode()
        scope = dict(scope)
        scope["headers"] = _with_content_length(scope.get("headers", []), len(new_body))
        scope.setdefault("state", {})["id_map"] = id_map

        if not id_map:
            await self.app(scope, _replay(new_body), send)
            return

        await self.app(scope, _replay(new_body), _decorating_send(send, id_map))


async def _read_body(receive: Any) -> bytes:
    chunks = []
    while True:
        message = await receive()
        if message["type"] != "http.request":
            break
        chunks.append(message.get("body", b""))
        if not message.get("more_body", False):
            break
    return b"".join(chunks)


def _replay(body: bytes) -> Any:
    """A ``receive`` callable that replays ``body`` to the downstream app."""
    sent = False

    async def receive() -> Dict[str, Any]:
        nonlocal sent
        if sent:
            return {"type": "http.disconnect"}
        sent = True
        return {"type": "http.request", "body": body, "more_body": False}

    return receive


def _with_content_length(headers: Any, length: int) -> list:
    out = [(k, v) for k, v in headers if k.lower() != b"content-length"]
    out.append((b"content-length", str(length).encode()))
    return out


def _decorating_send(send: Any, id_map: Dict[str, str]) -> Any:
    """Wrap ``send`` so a JSON object response gains ``idMap``.

    The rewritten payload changes length, so the buffered body is emitted with a
    corrected content-length rather than the handler's original.
    """
    import json

    state: Dict[str, Any] = {"start": None, "chunks": [], "json": False}

    async def send_wrapper(message: Dict[str, Any]) -> None:
        if message["type"] == "http.response.start":
            headers = message.get("headers", [])
            state["json"] = any(
                k.lower() == b"content-type" and b"application/json" in v.lower()
                for k, v in headers
            )
            if not state["json"]:
                await send(message)
                return
            state["start"] = message
            return

        if message["type"] == "http.response.body" and state["json"]:
            state["chunks"].append(message.get("body", b""))
            if message.get("more_body", False):
                return

            raw = b"".join(state["chunks"])
            body = raw
            try:
                payload = json.loads(raw)
                if isinstance(payload, dict) and "idMap" not in payload:
                    body = json.dumps({**payload, "idMap": id_map}).encode()
            except ValueError:
                pass

            start = dict(state["start"])
            start["headers"] = _with_content_length(start.get("headers", []), len(body))
            await send(start)
            await send({"type": "http.response.body", "body": body, "more_body": False})
            return

        await send(message)

    return send_wrapper


async def _send_json(send: Any, status: int, payload: Dict[str, Any]) -> None:
    import json

    body = json.dumps(payload).encode()
    await send(
        {
            "type": "http.response.start",
            "status": status,
            "headers": [
                (b"content-type", b"application/json"),
                (b"content-length", str(len(body)).encode()),
            ],
        }
    )
    await send({"type": "http.response.body", "body": body, "more_body": False})

File: packages/identity/src/graph-create.ts
// Graph create: `lid` in, `idMap` out.
//
// The problem: once clients stop supplying ids, a frontend building a customer
// plus its orders in one request can no longer express "this order belongs to
// THAT customer I am creating right now", and cannot learn the real ids on
// return.
//
// The shape: each node carries a document-scoped local id (`lid`) and
// references it as `"lid:<local>"`. Standardised prior art, not invention —
// JSON:API 1.1 `lid`, SCIM `bulkId` (RFC 7644 3.7.2), OData $batch Content-ID.
// Deliberately NOT the `id` field: putting a placeholder there would put `id`
// back in create bodies and degrade the rule to "id must be present but fake",
// which is unenforceable. `id` is never accepted, full stop.
//
// The trick: every id is minted UP FRONT, before any handler runs, and
// substituted into the body. Three consequences —
//   1. handlers receive a plain body with real ids and never learn `lid`
//      existed, so no route implements any of this;
//   2. `idMap` is just the allocation table, so nothing rewrites on the way out;
//   3. cycles resolve. A -> B and B -> A is unsolvable by post-substitution
//      (it needs a deferred UPDATE); with pre-allocation both ids exist before
//      either row is written.
//
// Policy: governance/identifier-standard.md 3-4.

import { EntityId, IdentityError } from './brand'
import { mintId } from './id'
import { EntityType, isEntityType } from './registry'

/** Reference token: `lid:<local>`. */
const LID_REF = /^lid:(.+)$/

/**
 * Permitted shape of a local id, matching the `LocalId` schema published in the
 * contract (1-64 chars). Enforced here because `lid` values are echoed back to
 * the caller as `idMap` KEYS — the one place client-controlled text crosses into
 * a response — and because a contract that declares a bound the implementation
 * does not enforce is not a bound at all.
 */
const LID_FORMAT = /^[A-Za-z0-9_.:-]{1,64}$/

export interface GraphCreateOptions {
  /**
   * Entity types this service OWNS. A `lid` node declaring anything else is
   * rejected: a lid graph is scoped to one service's aggregate boundary,
   * because a graph spanning services cannot be created atomically — there is
   * no distributed transaction. Cross-service references must be real,
   * already-existing ids. See governance/identifier-standard.md 4.
   */
  aggregate: ReadonlySet<EntityType> | readonly EntityType[]
  /** Node-count ceiling; an unbounded graph is a cheap amplification vector. */
  maxNodes?: number
  /** Nesting ceiling, guarding against deep or cyclic input structures. */
  maxDepth?: number
}

export interface GraphCreateResult {
  /** The body with every `lid` resolved to a minted id and `lid` removed. */
  body: unknown
  /** local id -> minted entity id, for every first-class entity created. */
  idMap: Record<string, string>
}

export type GraphErrorCode =
  | 'CLIENT_SUPPLIED_ID'
  | 'UNKNOWN_LID'
  | 'DUPLICATE_LID'
  | 'MALFORMED_LID'
  | 'MISSING_TYPE'
  | 'UNKNOWN_TYPE'
  | 'CROSS_AGGREGATE_LID'
  | 'GRAPH_TOO_LARGE'
  | 'GRAPH_TOO_DEEP'

export class GraphCreateError extends Error {
  readonly code: GraphErrorCode
  /** JSON-pointer-ish path to the offending node, for actionable 422s. */
  readonly path: string

  constructor(code: GraphErrorCode, path: string, message: string) {
    super(message)
    this.name = 'GraphCreateError'
    this.code = code
    this.path = path
  }
}

const DEFAULT_MAX_NODES = 500
const DEFAULT_MAX_DEPTH = 32

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

/**
 * Resolves a create-graph: validates it, mints an id per `lid` node, and
 * rewrites every `lid:` reference to the minted id.
 *
 * Framework-agnostic on purpose — the Express adapter below is a thin shell, so
 * this stays unit-testable without spinning up a server.
 *
 * Mutates nothing: returns a rewritten copy.
 */
export function resolveGraph(body: unknown, options: GraphCreateOptions): GraphCreateResult {
  const aggregate =
    options.aggregate instanceof Set ? options.aggregate : new Set(options.aggregate)
  const maxNodes = options.maxNodes ?? DEFAULT_MAX_NODES
  const maxDepth = options.maxDepth ?? DEFAULT_MAX_DEPTH

  // Pass 1 — validate, and mint an id for every lid node. Minting up front is
  // what makes reference cycles resolvable in pass 2.
  const idMap: Record<string, string> = {}
  let nodeCount = 0

  const collect = (value: unknown, path: string, depth: number): void => {
    if (depth > maxDepth) {
      throw new GraphCreateError('GRAPH_TOO_DEEP', path, `graph exceeds ${maxDepth} levels`)
    }
    if (Array.isArray(value)) {
      value.forEach((item, i) => collect(item, `${path}/${i}`, depth + 1))
      return
    }
    if (!isPlainObject(value)) return

    if ('id' in value) {
      throw new GraphCreateError(
        'CLIENT_SUPPLIED_ID',
        path,
        'ids are minted by the owning service and must not be supplied on create'
      )
    }

    if ('lid' in value) {
      const lid = value.lid
      if (typeof lid !== 'string' || !LID_FORMAT.test(lid)) {
        throw new GraphCreateError(
          'MALFORMED_LID',
          path,
          'lid must be 1-64 characters of [A-Za-z0-9_.:-]'
        )
      }
      if (Object.prototype.hasOwnProperty.call(idMap, lid)) {
        throw new GraphCreateError('DUPLICATE_LID', path, `lid ${JSON.stringify(lid)} is declared twice`)
      }
      if (++nodeCount > maxNodes) {
        throw new GraphCreateError('GRAPH_TOO_LARGE', path, `graph exceeds ${maxNodes} nodes`)
      }

      const type = value.type
      if (typeof type !== 'string') {
        throw new GraphCreateError('MISSING_TYPE', path, 'a lid node must declare its entity type')
      }
      if (!isEntityType(type)) {
        throw new GraphCreateError('UNKNOWN_TYPE', path, `unregistered entity type ${JSON.stringify(type)}`)
      }
      if (!aggregate.has(type)) {
        throw new GraphCreateError(
          'CROSS_AGGREGATE_LID',
          path,
          `${type} is owned by another service; reference it by its existing id instead of creating it here`
        )
      }

      idMap[lid] = mintId(type)
    }

    for (const [key, child] of Object.entries(value)) {
      collect(child, `${path}/${key}`, depth + 1)
    }
  }

  collect(body, '', 0)

  // Pass 2 — rewrite. Every id already exists, so order does not matter and
  // mutual references resolve without a second write.
  const rewrite = (value: unknown, path: string): unknown => {
    if (Array.isArray(value)) return value.map((item, i) => rewrite(item, `${path}/${i}`))
    if (typeof value === 'string') {
      const match = LID_REF.exec(value)
      if (!match) return value
      const target = match[1]
      const resolved = idMap[target]
      if (!resolved) {
        throw new GraphCreateError(
          'UNKNOWN_LID',
          path,
          `reference to undeclared lid ${JSON.stringify(target)}`
        )
      }
      return resolved
    }
    if (!isPlainObject(value)) return value

    const out: Record<string, unknown> = {}
    for (const [key, child] of Object.entries(value)) {
      if (key === 'lid') continue // document-scoped; never persisted
      out[key] = rewrite(child, `${path}/${key}`)
    }
    if (typeof value.lid === 'string') {
      // The server-minted id, injected where the handler expects to find it.
      out.id = idMap[value.lid]
    }
    return out
  }

  return { body: rewrite(body, ''), idMap }
}

// ---------------------------------------------------------------------------
// Express adapter
// ---------------------------------------------------------------------------

// Typed structurally rather than against @types/express on purpose: this repo
// pins @types/express inconsistently across workspaces (4.17.21 and ^5.0.0 both
// appear), and a hard dependency here would drag that conflict into every
// consumer. These shapes are satisfied by Express 4 and 5 alike.
export interface GraphRequestLike {
  method: string
  body?: unknown
}

export interface GraphResponseLike {
  locals: Record<string, unknown>
  status(code: number): GraphResponseLike
  json(payload: unknown): unknown
}

export type GraphNext = (err?: unknown) => void

const MUTATING_METHODS = new Set(['POST', 'PUT', 'PATCH'])

/**
 * Express middleware wrapping `resolveGraph`.
 *
 * Mount it ahead of the create routes. Handlers downstream see a body whose
 * ids are already minted and whose references are already resolved, and the
 * `idMap` is merged into the JSON response automatically — so a route opts in
 * to graph create by doing nothing at all.
 */
export function graphCreate(options: GraphCreateOptions) {
  return function graphCreateMiddleware(
    req: GraphRequestLike,
    res: GraphResponseLike,
    next: GraphNext
  ): void {
    if (!MUTATING_METHODS.has(req.method.toUpperCase()) || req.body === undefined) {
      next()
      return
    }

    let result: GraphCreateResult
    try {
      result = resolveGraph(req.body, options)
    } catch (err) {
      if (err instanceof GraphCreateError || err instanceof IdentityError) {
        res.status(422).json({
          error: 'unprocessable_entity',
          code: err.code,
          message: err.message,
          // Always present for a graph error, including the empty string that
          // denotes the document root — an omitted path reads as "location
          // unknown", which is exactly wrong for the root-level case.
          ...(err instanceof GraphCreateError ? { path: err.path } : {}),
        })
        return
      }
      next(err)
      return
    }

    req.body = result.body
    res.locals.idMap = result.idMap

    if (Object.keys(result.idMap).length > 0) {
      const originalJson = res.json.bind(res)
      res.json = (payload: unknown): unknown => {
        // Only decorate object payloads, and never clobber a handler that has
        // already produced its own idMap.
        if (isPlainObject(payload) && !('idMap' in payload)) {
          return originalJson({ ...payload, idMap: result.idMap })
        }
        return originalJson(payload)
      }
    }

    next()
  }
}

/** The minted ids for a request, for handlers that need them directly. */
export function idMapOf(res: GraphResponseLike): Record<string, string> {
  return (res.locals.idMap as Record<string, string> | undefined) ?? {}
}

/** Re-exported so consumers can type an id column without a second import. */
export type { EntityId }
